web-api/src/storage/elastic/filter.rs

```diff
@@ -70,34 +71,141 @@ impl Serialize for Terms<'_> {
     }
 }
 
+fn cmp_range(this: &Value, other: &Value) -> Option<Ordering> {
+    match (this, other) {
+        (Value::Number(this), Value::Number(value)) => {
+            // unwraps are always possible without feature `arbitrary_precision`
+            Some(this.as_f64().unwrap().total_cmp(&value.as_f64().unwrap()))
+        }
+        (Value::String(this), Value::String(value)) => {
+            // unwraps are always possible since range strings are valid dates
+            Some(
+                DateTime::parse_from_rfc3339(this)
+                    .unwrap()
+                    .cmp(&DateTime::parse_from_rfc3339(value).unwrap()),
+            )
+        }
+        // TODO: remove Option<_> and change to unreachable! once filters are validated against indexed properties schema
+        _ => None,
+        // _ => unreachable!(/* values are both numbers or dates */),
+    }
+}
+
 #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize)]
 #[serde(rename_all = "lowercase")]
-enum RangeOp {
+enum GreaterRangeOp {
     Gt,
     Gte,
+}
+
+#[derive(Debug)]
+struct GreaterRange<'a> {
+    operation: GreaterRangeOp,
+    value: &'a DocumentProperty,
+}
+
+impl GreaterRange<'_> {
+    fn and(&mut self, greater: Self) {
+        match cmp_range(self.value, greater.value) {
+            Some(Ordering::Less) => *self = greater,
+            Some(Ordering::Equal) => {
+                if self.operation == GreaterRangeOp::Gte && greater.operation == GreaterRangeOp::Gt
+                {
+                    self.operation = greater.operation;
+                }
+            }
+            Some(Ordering::Greater) | None => {}
+        }
+    }
+
+    fn or(&mut self, greater: Self) {
+        match cmp_range(self.value, greater.value) {
+            Some(Ordering::Less) | None => {}
+            Some(Ordering::Equal) => {
+                if self.operation == GreaterRangeOp::Gt && greater.operation == GreaterRangeOp::Gte
+                {
+                    self.operation = greater.operation;
+                }
+            }
+            Some(Ordering::Greater) => *self = greater,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize)]
+#[serde(rename_all = "lowercase")]
+enum LessRangeOp {
     Lt,
     Lte,
 }
 
+#[derive(Debug)]
+struct LessRange<'a> {
+    operation: LessRangeOp,
+    value: &'a DocumentProperty,
+}
+
+impl LessRange<'_> {
+    fn and(&mut self, less: Self) {
```

Comment:
:art: ? minor nit-pick by convention of implementing operator like function the parameter containing the other value used with the operator is most times called `other`.  This has the benefit or not carrying any potentially accident meaning. Like here `less` not necessary being less then self.